### DISCUSSION TITLE
[@types/three] Understanding interface changes

### DISCUSSION THREAD

**Discussion by EdwardMoyse** [DefinitelyTyped/DefinitelyTyped#69506] [Issues with a @types package] [answered]
This is mostly just for my education. My project has a [problem](https://github.com/HSF/phoenix/issues/658) upgrading to three v0.164 basically because we had this in our code:
```
  private textFont: Font = new Font(HelvetikerFont);
```

https://github.com/HSF/phoenix/blob/13cdc34bdde7800080d04da342e3d9b483df18ec/packages/phoenix-event-display/src/managers/three-manager/scene-manager.ts#L68

`FontData` was introduced in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69203, which breaks this.

However if I look at the library documentation for [FontLoader](https://threejs.org/docs/#examples/en/loaders/FontLoader.parse)
it has a different interface, namely `parse(json: Object)` instead of `parse(json: FontData)` in https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/three.

I'm just wondering if it is normal to have a more restrictive interface here than in the base library? I was *very* confused about what was going on, until I looked at `FontLoader.d.ts`

[a comment by an automated account is not shown]

**Comment by Methuselah96** [comment, marked as answer]
Hey @EdwardMoyse thanks for raising this issue!

> However if I look at the library documentation for [FontLoader](https://threejs.org/docs/#examples/en/loaders/FontLoader.parse)
it has a different interface, namely parse(json: Object) instead of parse(json: FontData) in https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/three.
>
> I'm just wondering if it is normal to have a more restrictive interface here than in the base library? I was very confused about what was going on, until I looked at FontLoader.d.ts

In short, yes. The documentation doesn't always fully spell out the types of `Object` in cases like this, but the TypeScript types are a bit more fleshed out for better type-safety.

The issue that you're having is actually that some of your glyphs don't have the `o` property. This is an oversight in the three.js types, and should be fixed by https://github.com/three-types/three-ts-types/pull/934 and released as part of @types/three@0.165.0.

Let me know if you have any other questions!

**Reply by EdwardMoyse** [reply]
Thanks for the explanation. I'll wait for 0.165 then.
